File: ios/BridgeInitialScreen.swift
import Foundation
import Octopus
import OctopusUI

/// Which native view the bridge mounts for a decoded `initialScreen` payload.
///
/// Every case but one is a native `OctopusInitialScreen` handed to `OctopusHomeScreen`. The JS
/// `profile` screen has **no counterpart in the native iOS enum** — iOS ships a separate
/// top-level `OctopusProfileScreen` view for it (Android pushes a `ProfileSummary` /
/// `CurrentUserProfileGraph` destination inside the same NavHost as the other bridge-mode
/// entries). So the decode result has to be able to say "mount a different view", not just
/// "which start screen". Same shape as the Flutter bridge's `BridgeInitialScreen`.
enum BridgeInitialScreen {
  /// Mount `OctopusHomeScreen` with this start screen.
  case home(OctopusInitialScreen)
  /// Mount `OctopusProfileScreen` for this member, or — `nil` — for the connected user's own
  /// (editable) profile.
  case profile(clientUserId: String?)
}

/// Decodes the flat `initialScreen` wire payload produced by the JS `normalizeInitialScreen`
/// (`src/internals/initialScreen.ts`) into the native view the bridge should mount.
///
/// A pure decoder: ids are forwarded exactly as they arrived — the JS producer already trimmed
/// and validated them, and a decoder that also normalized would hide a producer divergence
/// between the two platforms instead of ruling it out. Unknown / structurally impossible
/// payloads (blank required id, activity without exactly one member id) fold to `.mainFeed`
/// with a log rather than crashing the host app.
///
/// The `createPost` type is deliberately **not** handled here: `openUI` decodes it itself so an
/// invalid prefill can reject the promise with the same error codes as
/// `navigateToOctopusCreatePost`, which a fold-to-mainFeed decoder cannot do.
func decodeBridgeInitialScreen(_ map: [String: Any]) -> BridgeInitialScreen {
  func string(_ key: String) -> String? {
    (map[key] as? String)?.nilIfBlank
  }
  switch map["type"] as? String {
  case "mainFeed", nil:
    return .home(.mainFeed)
  case "post":
    guard let postId = string("postId") else {
      NSLog("[OctopusReactNativeSdk] initialScreen.post: missing/blank postId — falling back to mainFeed")
      return .home(.mainFeed)
    }
    return .home(.post(.init(postId: postId)))
  case "group":
    guard let groupId = string("groupId") else {
      NSLog("[OctopusReactNativeSdk] initialScreen.group: missing/blank groupId — falling back to mainFeed")
      return .home(.mainFeed)
    }
    return .home(.group(.init(groupId: groupId)))
  case "activity":
    let profileId = string("profileId")
    let clientUserId = string("clientUserId")
    if let clientUserId, profileId == nil {
      return .home(.activity(.init(clientUserId: clientUserId)))
    }
    if let profileId, clientUserId == nil {
      return .home(.activity(.init(profileId: profileId)))
    }
    NSLog("[OctopusReactNativeSdk] initialScreen.activity: needs exactly one member id — falling back to mainFeed")
    return .home(.mainFeed)
  case "profile":
    // An absent/blank id means the connected user's own profile —
    // `OctopusProfileScreen(clientUserId: nil)`.
    return .profile(clientUserId: string("clientUserId"))
  default:
    NSLog("[OctopusReactNativeSdk] initialScreen: unknown type=\(map["type"] ?? "nil") — falling back to mainFeed")
    return .home(.mainFeed)
  }
}

private extension String {
  /// `nil` when there is nothing but whitespace here, otherwise the string **unchanged** — the
  /// same "missing/blank" test the Android decoder applies. It deliberately does not return the
  /// trimmed value: the JS producer trims, and a decoder that also normalized would hide a
  /// producer divergence between the platforms instead of ruling it out.
  var nilIfBlank: String? {
    trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
  }
}

File: android/src/main/java/com/octopuscommunity/octopusreactnativesdk/BridgeInitialScreen.kt
package com.octopuscommunity.octopusreactnativesdk

import android.util.Log

/**
 * The decoded `initialScreen` payload carried from [OctopusUIController.openUI] to
 * [OctopusContent] through [OctopusActivity]'s Intent extras — and, on the embedded path, from
 * the `initialScreen` prop of [OctopusUIViewManager] straight into [OctopusContent].
 *
 * Only the screens that navigate on top of the home NavHost are modeled here; `mainFeed` is the
 * absence of an initial screen (`null`), and `createPost` rides the existing
 * `EXTRA_CREATE_POST_*` extras into [OctopusContent]'s `createPostInfo` parameter, so neither
 * needs a case. Ids arrive already trimmed and validated by the JS producer
 * (`normalizeInitialScreen`); the decoders here only reject structurally impossible payloads
 * (blank required id, activity without exactly one id), folding to the main feed with a warning
 * rather than crashing the host app.
 */
internal sealed interface BridgeInitialScreen {
  data class Post(val postId: String) : BridgeInitialScreen

  data class Group(val groupId: String) : BridgeInitialScreen

  /** Exactly one of [profileId] / [clientUserId] is non-null. */
  data class MemberActivity(val profileId: String?, val clientUserId: String?) :
    BridgeInitialScreen

  /** [clientUserId] `null` means the connected user's own, editable profile. */
  data class Profile(val clientUserId: String?) : BridgeInitialScreen
}

/**
 * Decodes a `type` plus a string accessor into a [BridgeInitialScreen]. `mainFeed`, `createPost`
 * (handled through `createPostInfo`) and an absent type all decode to `null`.
 *
 * Source-agnostic on purpose: the fullscreen path reads Intent extras and the embedded path reads
 * a `ReadableMap` prop, so each supplies its own [string] accessor and both get the exact same
 * folding rules. Ids are forwarded verbatim — the JS producer already trimmed and validated them.
 *
 * @param string reads one payload field; `null` when absent.
 * @param tag log tag identifying the calling path.
 */
internal fun decodeBridgeInitialScreen(
  type: String?,
  tag: String,
  string: (String) -> String?
): BridgeInitialScreen? {
  fun foldMalformed(reason: String): BridgeInitialScreen? {
    Log.w(tag, "Malformed initialScreen ($reason) — opening the main feed")
    return null
  }
  return when (type) {
    null, "mainFeed", "createPost" -> null
    "post" -> {
      val postId = string("postId")
      if (postId.isNullOrBlank()) foldMalformed("post without postId")
      else BridgeInitialScreen.Post(postId)
    }
    "group" -> {
      val groupId = string("groupId")
      if (groupId.isNullOrBlank()) foldMalformed("group without groupId")
      else BridgeInitialScreen.Group(groupId)
    }
    "activity" -> {
      val profileId = string("profileId")?.takeUnless { it.isBlank() }
      val clientUserId = string("clientUserId")?.takeUnless { it.isBlank() }
      if ((profileId != null) == (clientUserId != null)) {
        foldMalformed("activity without exactly one member id")
      } else {
        BridgeInitialScreen.MemberActivity(profileId = profileId, clientUserId = clientUserId)
      }
    }
    "profile" -> BridgeInitialScreen.Profile(
      clientUserId = string("clientUserId")?.takeUnless { it.isBlank() }
    )
    else -> foldMalformed("unknown type \"$type\"")
  }
}
